File: erlang/bob/src/bob.erl
-module(bob).

-export([response/1]).

response(Str) ->
    case string:trim(Str) of
        [] -> "Fine. Be that way!";
        Input -> response(Input, Str)
    end.

response(Input, Str) ->
    Question = string:equal(string:substr(Input, length(Input)), "?"),
    Upper = string:uppercase(Str),
    Scream = (Upper == Str) and (string:lowercase(Str) /= Upper),
    case {Scream, Question} of
        {true, true} -> "Calm down, I know what I'm doing!";
        {false, true} -> "Sure.";
        {true, false} -> "Whoa, chill out!";
        {false, false} -> "Whatever."
    end.

File: elixir/bob/lib/bob.ex
defmodule Bob do
  import String, only: [trim: 1, ends_with?: 2, upcase: 1, downcase: 1]

  def hey(input) do
    cond do
      anything?(input) -> "Fine. Be that way!"
      scream?(input) && question?(input) -> "Calm down, I know what I'm doing!"
      question?(input) -> "Sure."
      scream?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp anything?(input), do: trim(input) == ""
  defp question?(input), do: ends_with?(trim(input), "?")

  defp scream?(input),
    do: upcase(input) == input && downcase(input) !== upcase(input)
end
